lib/nrf_modem_lib/trace_medium.c

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <string.h>
+#include <zephyr/zephyr.h>
+#include <zephyr/logging/log.h>
+#include <nrf_modem.h>
+#include <modem/trace_medium.h>
+
+LOG_MODULE_REGISTER(trace_medium, CONFIG_NRF_MODEM_LIB_LOG_LEVEL);
+
+static struct trace_medium_entry *medium = NULL;
+
+bool trace_medium_init(void)
+{
+	int err;
+
+	/* Attempt to reinitialize medium */
+	if (medium != NULL) {
+		err = medium->init();
+		if (err == 0 || err == EALREADY) {
+			return true;
+		} else {
+			LOG_ERR("Failed to reinit trace medium %s, err %d", medium->name, err);
+			medium = NULL;
+		}
+	}
+
+	/* Find and attempt to initialize a medium */
+	STRUCT_SECTION_FOREACH(trace_medium_entry, m) {
+		if (m->init == NULL || m->deinit == NULL || m->write == NULL) {
```

Comment:
As long as we only want one trace medium we should assert on the size of the iterable section and make sure the first (and only) element is valid, and log and return an error if that is not the case. Here we have the potential of configuring multiple trace mediums, where we simply chooses whatever is compiled at the start of the list.

Reply:
If we had only one trace medium, we won't need any of these macros and `NULL` checks

Reply:
Actually, this whole file as it is would disappear, and we would just have:
`trace_medium_rtt.c` with an implementation of `trace_medium_init`, `_write` etc based on RTT and
`trace_medium_uart.c` with an implementation of `trace_medium_init`, etc based on UART

Reply:
Yes, have already removed it locally.